### PR TITLE
Obj loader

### PR DESCRIPTION
**Pull request by Korijn**
Getting some weird artefacts... not sure why yet:

image

Run like this: `~/dev/radiant$ PYTHONPATH=$PWD python examples/dragon_scene.py`

### CODE REVIEW

File: radiant/loaders.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    if winding_order.lower() != 'ccw':
+        f = flip_winding_order(f)
+
+    v = np.array(v, dtype='f4')[f[:, :, 0].ravel()]
+    vt = np.array(vt, dtype='f4')[f[:, :, 1].ravel()]
+    vn = np.array(vn, dtype='f4')[f[:, :, 2].ravel()]
+
+    # NOTE: we could identify unique vertices in f and shave off memory requirements. we could.
+
+    return Geometry({'pos': v, 'uv': vt, 'normal': vn}, primitive=Primitives.TRIANGLES)
```
Comment (Korijn) [inline radiant/loaders.py]: winding order should be a property of Geometry

**Comment by cprogrammer1994** [comment]
Just a reply for the "weird" output...

Do you have depth test enabled?
You can make a comparison using the ModernGL.ext.obj
There is no randomness while loading OBJ files. I guess you can simply binary compare the results.

File: radiant/renderers/moderngl.py [hunk 2]
```diff
@@ -27,8 +27,10 @@ def visit(node):
 
         # get going
         self.ctx.enable(ModernGL.DEPTH_TEST)
```
Comment (Korijn) [inline radiant/renderers/moderngl.py]: @cprogrammer1994 I've enabled DEPTH_TEST here, and CULL_FACE too.

File: radiant/loaders.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        elif line.startswith("vn "):
+            vn.append([float(x) for x in line[3:].split()])
+        elif line.startswith("f "):
+            f.append([[int(x) for x in fv.split('/')] for fv in line[2:].split()])
+
+    # now produce the full buffers
+    f = np.array(f, dtype='i4')
+    f[f > 0] -= 1
+    if winding_order.lower() != 'ccw':
+        f = flip_winding_order(f)
```
Comment (Korijn) [inline radiant/loaders.py]: @cprogrammer1994 do you know if the dragon file has its triangles defined in clockwise or counter-clockwise order?

Comment (cprogrammer1994) [inline radiant/loaders.py]: In my dragon example I use face culling and the default OpenGL winding is CCW for the front-face.

dragon.obj -> front-face = ccw

Comment (Korijn) [inline radiant/loaders.py]: okay, well, that's not the issue here then...

```python
with open("examples/resources/dragon.obj", mode="r") as fh:
    dragon_geometry = radiant.load_obj(fh, winding_order='ccw')
```

File: radiant/renderers/moderngl.py (hunk 2 above)
Comment (Korijn) [inline radiant/renderers/moderngl.py]: It does look like the depth_test is failing though... not sure what's going on there. should I explicitly clear the depth buffer?

**Comment by cprogrammer1994** [comment]
image

I hope it is not a ModernGL issue :))

**Comment by cprogrammer1994** [comment]
what OS do you use? (+ download link?)

**Comment by Korijn** [comment]
Is that the output on your machine? Wow!

I'm running [Ubuntu GNOME 17.04](https://wiki.ubuntu.com/UbuntuGNOME/GetUbuntuGNOME).

**Comment by cprogrammer1994** [comment]
To avoid depth fighting change:

```
fov=45.0, aspect=4.0/3.0, near=0.1, far=1000.0
```

to

```
fov=45.0, aspect=4.0/3.0, near=0.5, far=200.0
```

Let see if the pixels change.

---

It can also happen if the fragment shader processes some `NaN` values the pixels get discarded.

File: radiant/renderers/moderngl.py (hunk 2 above)
Comment (cprogrammer1994) [inline radiant/renderers/moderngl.py]: clear should clear both. [proof](https://github.com/cprogrammer1994/ModernGL/blob/master/src/Context.cpp#L113)

**Comment by Korijn** [comment]
PR is done. Artefacts are being caused by hardware/driver glitches on my machine... haven't been able to resolve that yet.

**Comment by cprogrammer1994** [comment]
I thought that your depth buffer was 8 bit.
This is deprecated but you can try to query: `GL_DEPTH_BITS`

**Comment by Korijn** [comment]
See here for the fix that I found later: https://github.com/Korijn/radiant/pull/16

**Pull request merged**